exe/Main.hs

```diff
@@ -188,22 +192,256 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+        cached_targets <- concatMapM (fmap fst . new_cache) old_deps
+        modifyVar_ fileToFlags $ \var -> do
+            pure $ Map.insert hieYaml (HM.fromList (cs ++ cached_targets)) var
+
+        return res
+
+    lock <- newLock
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    sessionOpts <- return $ \(hieYaml, file) -> do
+
+
+        fm <- readVar fileToFlags
+        let mv = Map.lookup hieYaml fm
+        let v = fromMaybe HM.empty mv
+        cfp <- liftIO $ canonicalizePath file
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+          Just (_, old_di) -> do
+            deps_ok <- checkDependencyInfo old_di
+            unless deps_ok $ do
+              modifyVar_ fileToFlags (const (return Map.empty))
+              -- Keep the same name cache
+              modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+          Nothing -> return ()
+        -- We sort so exact matches come first.
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just opts -> do
+                --putStrLn $ "Cached component of " <> show file
+                pure (fst opts)
+            Nothing-> do
+                finished_barrier <- newBarrier
+                -- fork a new thread here which won't be killed by shake
+                -- throwing an async exception
+                void $ forkIO $ do
+                  putStrLn $ "Consulting the cradle for " <> show file
+                  cradle <- maybe (loadImplicitCradle $ addTrailingPathSeparator dir) loadCradle hieYaml
+                  opts <- cradleToSessionOpts cradle cfp
+                  print opts
+                  res <- fst <$> session (hieYaml, toNormalizedFilePath' cfp, opts)
+                  signalBarrier finished_barrier res
+                waitBarrier finished_barrier
+
+    dummyAs <- async $ return (error "Uninitialised")
+    runningCradle <- newIORef dummyAs
+    -- The main function which gets options for a file. We only want one of these running
+    -- at a time.
+    let getOptions file = do
+            hieYaml <- cradleLoc file
+            sessionOpts (hieYaml, file)
+    -- The lock is on the `runningCradle` resource
+    return $ \file -> liftIO $ withLock lock $ do
+        as <- readIORef runningCradle
+        finished <- poll as
+        case finished of
+            Just {} -> do
+                as <- async $ getOptions file
+                writeIORef runningCradle as
+                wait as
+             -- If it's not finished then wait and then get options, this could of course be killed still
+            Nothing -> do
+                _ <- wait as
+                getOptions file
+
+
+
+
+
+checkDependencyInfo :: Map.Map FilePath (Maybe UTCTime) -> IO Bool
+checkDependencyInfo old_di = do
+  di <- getDependencyInfo (Map.keys old_di)
+  return (di == old_di)
+
+
+
+getDependencyInfo :: [FilePath] -> IO (Map.Map FilePath (Maybe UTCTime))
```

Comment:
This looks like we are reimplementing file modification checking instead of piggy backing on the one we have for source files. I think the cradle setup is sufficiently special that I don’t mind this if reusing it is difficult but I’d like to have a comment that explains that this is only for cradle dependencies and ideally also an explanation why it is hard to reuse the one we already have.